core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableLimitSort.java

```diff
@@ -0,0 +1,146 @@
+  }
+
+  /** Creates an EnumerableLimitSort. */
+  public static EnumerableLimitSort create(
+      RelNode input,
+      RelCollation collation,
+      RexNode offset,
+      RexNode fetch) {
+    final RelOptCluster cluster = input.getCluster();
+    final RelTraitSet traitSet = cluster.traitSetOf(EnumerableConvention.INSTANCE).replace(
+        collation);
+    return new EnumerableLimitSort(cluster, traitSet, input, collation, offset, fetch);
+  }
+
+  @Override public EnumerableLimitSort copy(
+      RelTraitSet traitSet,
+      RelNode newInput,
+      RelCollation newCollation,
+      RexNode offset,
+      RexNode fetch) {
+    return new EnumerableLimitSort(
+        this.getCluster(),
+        traitSet,
+        newInput,
+        newCollation,
+        offset,
+        fetch);
+  }
+
+  @Override public Result implement(EnumerableRelImplementor implementor, Prefer pref) {
+    final BlockBuilder builder = new BlockBuilder();
+    final EnumerableRel child = (EnumerableRel) this.getInput();
+    final Result result = implementor.visitChild(this, 0, child, pref);
+    final PhysType physType = PhysTypeImpl.of(
+        implementor.getTypeFactory(),
+        this.getRowType(),
+        result.format);
+    Expression childExp = builder.append("child", result.block);
+
+    PhysType inputPhysType = result.physType;
+    final Pair<Expression, Expression> pair =
+        inputPhysType.generateCollationKey(this.collation.getFieldCollations());
+
+    Expression fetchVal;
+    if (this.fetch == null) {
+      fetchVal = Expressions.constant(Integer.valueOf(Integer.MAX_VALUE));
+    } else {
+      fetchVal = getExpression(this.fetch);
+    }
+
+    Expression offsetVal = this.offset == null ? Expressions.constant(Integer.valueOf(0))
+        : getExpression(this.offset);
+
+    builder.add(
+        Expressions.return_(
+            null, Expressions.call(
+                BuiltInMethod.ORDER_BY_WITH_FETCH_AND_OFFSET.method, Expressions.list(
+                    childExp,
+                    builder.append("keySelector", pair.left))
+                    .appendIfNotNull(builder.appendIfNotNull("comparator", pair.right))
+                    .appendIfNotNull(
+                        builder.appendIfNotNull("offset",
+                            Expressions.constant(offsetVal)))
+                    .appendIfNotNull(
+                        builder.appendIfNotNull("fetch",
+                            Expressions.constant(fetchVal)))
+            )));
+    return implementor.result(physType, builder.toBlock());
+  }
+
+  @Override public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+    final double rowCount = mq.getRowCount(this.input).doubleValue();
+    double toSort = this.fetch == null ? rowCount : RexLiteral.intValue(this.fetch);
+    if (this.offset != null) {
+      toSort += RexLiteral.intValue(this.offset);
+    }
+    toSort = Math.min(rowCount, toSort);
```

Comment:
Why not use `max`?

Reply:
If there are only rowCount rows, we only need to sort that many. E.g., if fetch is 100, but there are only 10 rows in the input, we only need to do lookups for the 10 rows.